File: Problems/0652. (M) Find Duplicate Subtrees/Java/solution.java
class Solution {
  Set<String> seen = new HashSet();
  Map<String,TreeNode> ret = new HashMap();

  public String recurse(TreeNode node) {
    // get string representations of subtrees recursively
    String l_str = (node.left != null) ? recurse(node.left) : "n";
    String r_str = (node.right != null) ? recurse(node.right) : "n";
    String in_str = l_str + Integer.toString(node.val) + r_str + ".";
    // no duplicate of subtree encountered previously
    if (!this.seen.contains(in_str))
      this.seen.add(in_str);
    // duplicate not yet added to result pool
    else if (!this.ret.containsKey(in_str))
      this.ret.put(in_str, node);

    return in_str;
  }
  public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
    // recursive DFS with hashing

    recurse(root);

    // return values only
    return new ArrayList<TreeNode>(this.ret.values());
  }
}

File: Problems/0652. (M) Find Duplicate Subtrees/Python3/solution.py
class Solution:
  def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
    # recursive DFS with hashing

    self.seen = set()
    self.ret = {}

    def recurse(node):
      # get string representations of subtrees recursively
      l_str = recurse(node.left) if node.left else 'n'
      r_str = recurse(node.right) if node.right else 'n'
      in_str = l_str + str(node.val) + r_str + '.'
      # no previous copy of tree rooted at node
      if in_str not in self.seen:
        self.seen.add(in_str)
      # duplicate not yet added to result pool
      elif in_str not in self.ret:
        self.ret[in_str] = node
      return in_str

    recurse(root)

    # return values only
    return self.ret.values()

